ddtrace/internal/processor/trace.py

```diff
@@ -332,40 +332,41 @@ def on_span_finish(self, span):
                     break
 
 
-class PeerServiceProcessor(SpanProcessor):
+class PeerServiceProcessor(TraceProcessor):
     def __init__(self, peer_service_config):
         self._config = peer_service_config
         self._set_defaults_enabled = self._config.set_defaults_enabled
+        self._mapping = self._config.peer_service_mapping
 
-    def on_span_start(self, span):
-        """
-        We don't do anything on span start
-        """
-        pass
-
-    def on_span_finish(self, span):
-        if self._set_defaults_enabled:
-            self._set_span_tag(span)
-        self._remap_peer_service(span)  # Remap regardless of whether defaults are enabled
-
-    def _set_span_tag(self, span):
-        if span.get_tag(self._config.tag_name):  # If the tag already exists, assume it is user generated
-            span.set_tag_str(self._config.source_tag_name, self._config.tag_name)
+    def process_trace(self, trace):
+        if not trace:
             return
 
-        if span.get_tag(SPAN_KIND) not in self._config.enabled_span_kinds:
-            return
+        traces_to_process = []
+        if not self._set_defaults_enabled:
+            traces_to_process = filter(lambda x: x.get_tag(self._config.tag_name), trace)
+        else:
+            traces_to_process = filter(
+                lambda x: x.get_tag(self._config.tag_name) or x.get_tag(SPAN_KIND) in self._config.enabled_span_kinds,
+                trace,
+            )
+        any(map(lambda x: self._update_peer_service_tags(x), traces_to_process))
```

Comment:
I find this code really hard to follow. I don't see why this is easier/better than just:

```python
for span in trace:
    if span_shouldnt_be_processed: continue
    
    # process span
```

It also seems like we are probably calling `get_tag(self._config.tag_name)` one more time than is required (ok, maybe that is being pedantic/optimizing too much, but calling it out)

Reply:
we should remove the usage of `any()` here, it doesn't help legibility.

`any()` is used to iterate over the `map()` because `map()` gives us a `map object` and won't actually iterate over `traces_to_process` until we iterate over the `map object`.

but `any()` will stop after the first successful (truthy) response from `_update_peer_service_tags`. It works here because `_update_peer_service_tags` doesn't return anything (so `None)` which is falsey. But if someone updates `_update_peer_service_tags` to return something truthy in the future, this might start to fail silently without anyone knowing why.

`list()` would be better here instead of `any()` (also comments on why)


I also think this means we are iterating over spans multiple times? Once to filter, and then once again over the spans that need processing?